### PR TITLE
Fix: Add missing iOS contexts

### LINKED ISSUE #759: missing fallback userId
According to the document, the SDK uses a random `installationId` as a fallback for the user ID.

https://github.com/getsentry/sentry-docs/pull/4201
https://docs.sentry.io/platforms/flutter/enriching-events/identify-user/

However, I could not find an implementation of this and the reported issue seems to be tagged as `Unknown User`.

image

Is this behavior unimplemented? Or if I've missed something, please let me know.

### PR DESCRIPTION
**Pull request by marandaneto**
## :scroll: Description

- Add additional event data taken from native iOS bridge.

## :bulb: Motivation and Context
Closes https://github.com/getsentry/sentry-dart/issues/759

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: flutter/ios/Classes/SentryFlutterPluginApple.swift [hunk 1]
```diff
@@ -73,17 +73,51 @@ public class SentryFlutterPluginApple: NSObject, FlutterPlugin {
...
 
+            if let tags = serializedScope["tags"] as? [String: String] {
+                infos["tags"] = tags
+            }
+            if let extra = serializedScope["extra"] as? [String: Any] {
+                infos["extra"] = extra
+            }
+            if let user = serializedScope["user"] as? [String: Any] {
+                infos["user"] = user
+            }
```
Comment (marandaneto) [inline flutter/ios/Classes/SentryFlutterPluginApple.swift:99]: @brustolin the `user` here isn't set, so where should I read the installation id from sentry-cocoa?

Comment (marandaneto) [reply flutter/ios/Classes/SentryFlutterPluginApple.swift:99]: https://github.com/getsentry/sentry-cocoa/pull/1680

File: flutter/lib/src/default_integrations.dart [hunk 2]
```diff
@@ -195,15 +196,95 @@ class _LoadContextsIntegrationEventProcessor extends EventProcessor {
         event = event.copyWith(contexts: eventContexts);
       }
 
-      if (infos['integrations'] != null) {
-        final integrations = List<String>.from(infos['integrations'] as List);
+      final tagsMap = infos['tags'];
```
Comment (marandaneto) [inline flutter/lib/src/default_integrations.dart]: tags, extra and fingerprint set on Dart have precedence over the ones set in the native layer if the key exists on both sides

File: flutter/lib/src/default_integrations.dart [hunk 3]
```diff
@@ -195,15 +196,95 @@ class _LoadContextsIntegrationEventProcessor extends EventProcessor {
...
+        }
+        event = event.copyWith(fingerprint: eventFingerprints);
+      }
+
+      final levelString = infos['level'];
+      if (event.level == null && levelString != null) {
+        event = event.copyWith(level: SentryLevel.fromName(levelString));
+      }
+
+      final breadcrumbsList = infos['breadcrumbs'];
```
Comment (marandaneto) [inline flutter/lib/src/default_integrations.dart]: We need to be sure that we don't duplicate the crumbs, and the timeline of the crumbs should make sense.

File: dart/lib/src/protocol/sentry_user.dart [hunk 4]
```diff
@@ -63,12 +63,18 @@ class SentryUser {
 
   /// Deserializes a [SentryUser] from JSON [Map].
   factory SentryUser.fromJson(Map<String, dynamic> json) {
+    Map<String, dynamic>? extras;
+    if (json['extras'] != null) {
+      extras = Map<String, dynamic>.from(json['extras'] as Map);
+    } else {
+      extras = null;
+    }
```
Comment (marandaneto) [inline dart/lib/src/protocol/sentry_user.dart]:
```suggestion
    var extras = json['extras'];
    if (extras != null) {
      extras = Map<String, dynamic>.from(extras as Map);
    }
```

I think that would be cleaner

File: flutter/lib/src/default_integrations.dart [hunk 5]
```diff
@@ -195,14 +196,87 @@ class _LoadContextsIntegrationEventProcessor extends EventProcessor {
         event = event.copyWith(contexts: eventContexts);
       }
 
+      final tagsMap = infos['tags'];
+      if (tagsMap != null) {
```
Comment (marandaneto) [inline flutter/lib/src/default_integrations.dart]:
```suggestion
      if (tagsMap?.isNotEmpty ?? false) {
```

That would spare some code to be run, same for other fields.

File: flutter/lib/src/default_integrations.dart [hunk 6]
```diff
@@ -195,14 +196,87 @@ class _LoadContextsIntegrationEventProcessor extends EventProcessor {
...
       final userMap = infos['user'];
       if (event.user == null && userMap != null) {
         final user = Map<String, dynamic>.from(userMap as Map);
         event = event.copyWith(user: SentryUser.fromJson(user));
       }
 
-      if (infos['integrations'] != null) {
-        final integrations = List<String>.from(infos['integrations'] as List);
+      final distString = infos['dist'];
+      if (event.dist == null && distString != null) {
```
Comment (marandaneto) [inline flutter/lib/src/default_integrations.dart:232]:
```suggestion
      if (event.dist == null && distString?.isNotEmpty ?? false) {
```

Same for other fields

File: flutter/lib/src/default_integrations.dart (hunk 3 above)
Comment (denrase) [reply flutter/lib/src/default_integrations.dart]: Right now the breadcrumbs form the native bridge are merged with the hybrid ones and sorted by date, is there anything else we need to check?

**Comment by marandaneto** [comment]
@denrase Some tests are failing on CI.

**Comment by marandaneto** [comment]
Screenshot 2022-04-26 at 14 08 39

The navigation and started breadcrumb are not sorted out correctly based on timestamp

**Comment by denrase** [comment]
@marandaneto Hmm, we just do a simple sort on the timestamps. Just tested this and it looks ok to me. 🤔 

Bildschirmfoto 2022-04-27 um 14 29 53

**Comment by marandaneto** [comment]
> @marandaneto Hmm, we just do a simple sort on the timestamps. Just tested this and it looks ok to me. 🤔
> 
> Bildschirmfoto 2022-04-27 um 14 29 53

I could not reproduce it again, so let's ship it.

**Comment by marandaneto** [comment]
@denrase approve it please so I can merge it, since I opened the PR, I cannot approve myself.

**Review by denrase — APPROVED** [review]

**Comment by denrase** [comment]
@marandaneto Sry about the delay!

**Pull request merged**